### PR TITLE
Added vector's support for min, max and clamp functions

### LINKED ISSUE #13926: Allow clamping for more than just floats/ints
While doing a tutorial from KidsCanCode, it dawned upon me that the `clamp()` function allows for floats/ints to be clamped, but not other variable types like `Vector2`s, `Vector3`s, or `Color`s that one might want to clamp to a range. What I'm proposing is to allow just that, as I don't imagine it would be too hard to implement and it would allow for code to be condensed a little more.

For example, for a `Vector2` named `position`, this functionality would allow this code snippet:
```gdscript
position.x = clamp(position.x, 0, get_viewport_rect().size.x)
position.y = clamp(position.y, 0, get_viewport_rect().size.y)
```
to be simplified to this:
```gdscript
position = clamp(position, Vector2(), get_viewport_rect().size)
```
(You can see how this might be useful for a `Color`, where there might be a lot more values to clamp.)

Ideally, I would like to see `clamp()` allowed for `Vector2`, `Vector3`, and `Color`, at the least. (Though other types similar in nature to those three might be nice to allow clamping for too!)

Hope this isn't too much to ask for. :)

### PR DESCRIPTION
**Pull request by Chaosus**
By my experience - this really useful in many scenarious for those functions.

Fix #13926 (dont know if its so useful for Color's so I've didnt touches it), partially adresses #8071. Added for GDScript, Mono, VisualScript.

### CODE REVIEW

**Comment by aaronfranke** [comment]
Instead of `clamp(min, max, val)` I think it would be better to have it as `val.clamp(min, max)`. This would be in line with other Vector2/3 methods and would avoid the arguments being `Variant` which @karroffel expressed concern about due to performance.

**Comment by Calinou** [comment]
@aaronfranke I'd be in favor of this too. Also, we already have `Vector2.clamped()`, but not `Vector3.clamped()`.

File: core/typedefs.h [hunk 1]
```diff
@@ -126,14 +126,38 @@ T *_nullptr() {
...
 #endif
 
+#ifndef CLAMP2
+#define CLAMP2(m_a, m_min, m_max) (Vector2(CLAMP(m_a.x, m_min.x, m_max.x), CLAMP(m_a.y, m_min.y, m_max.y)))
+#endif
+
+#ifndef CLAMP3
+#define CLAMP3(m_a, m_min, m_max) (Vector3(CLAMP(m_a.x, m_min.x, m_max.x), CLAMP(m_a.y, m_min.y, m_max.y), CLAMP(m_a.z, m_min.z, m_max.z)))
+#endif
+
```
Comment (neikeq) [inline core/typedefs.h]: Why are we not using templates instead?

Comment (Chaosus) [reply core/typedefs.h]: Because it leads to ambiguity problems in a lot of places.

**Comment by Chaosus** [comment]
@Calinou, @aaronfranke ok, I've fixed clamp to be local function in Vector2/3, also added them to GDNative

File: core/typedefs.h (hunk 1 above)
Comment (Chaosus) [reply core/typedefs.h]: I can try with inline overloading however..

Comment (neikeq) [reply core/typedefs.h]: Do you have an example I can use to understand this?

Comment (Chaosus) [reply core/typedefs.h]: Well actually Vector2 and Vector3 are not defined in this file - so we need to push the include to this file (which may be undesired), I've tried to use templates and there is a result:
So I've tried to push this code

```
template <class T>
inline T MIN(T a, T b) {
	return (a < b) ? a : b;
}

template <class T>
inline T MIN2(const T &a, const T &b) {
	return T(MIN(a.x, b.x), MIN(a.y, b.y));
}

template <class T>
inline T MIN3(const T &a, const T &b) {
	return T(MIN(a.x, b.x), MIN(a.y, b.y), MIN(a.z, b.z));
}

template <class T>
inline T MAX(T a, T b) {
	return (a > b) ? a : b;
}

template <class T>
inline T MAX2(const T &a, const T &b) {
	return T(MAX(a.x, b.x), MAX(a.y, b.y));
}

template <class T>
inline T MAX3(const T &a, const T &b) {
	return T(MAX(a.x, b.x), MAX(a.y, b.y), MAX(a.z, b.z));
}

template <class T>
inline T CLAMP(T value, T min, T max) {
	return (value < min) ? min : ((value > max) ? max : value);
}

template <class T>
inline T CLAMP2(const T &value, const T &min, const T &max) {
	return T(clamp(value.x, min.x, max.x), clamp(value.y, min.y, max.y));
}

template <class T>
inline T CLAMP3(const T &value, const T &min, const T &max) {
	return T(clamp(value.x, min.x, max.x), clamp(value.y, min.y, max.y), clamp(value.z, min.z, max.z));
}
```
to typedefs.h

image

Sorry for non-english output..

**Comment by Calinou** [comment]
@Chaosus We already have `Vector2.clamped()`, so Vector3 should use `Vector3.clamped()` instead. There's no need for a `Vector2.clamp()` too :slightly_smiling_face:

**Comment by Chaosus** [comment]
Im lost my patience, so closing this pr..

**Pull request closed without merging**